### ISSUE TITLE
Transpose and lazy signals

### ISSUE THREAD

**Issue by magnunor** [hyperspy/hyperspy#1982]
Transposing lazy signals currently leads to very non-optimal chunking:

```python
import dask.array as da
from hyperspy._signals.signal2d import LazySignal2D

data = da.random.random((50, 50, 100, 100), chunks=(10, 10, 20, 20))
s = LazySignal2D(data)
s.data
dask.array<da.random.random_sample, shape=(50, 50, 100, 100), dtype=float64, chunksize=(10, 10, 20, 20)>

s.T.data
dask.array<rechunk-merge, shape=(100, 100, 50, 50), dtype=float64, chunksize=(50, 100, 50, 50)>

s1 = s.deepcopy()
s1.data = da.swapaxes(da.swapaxes(s.data, 0, 2), 1, 3)
s1.data
dask.array<swapaxes, shape=(100, 100, 50, 50), dtype=float64, chunksize=(20, 20, 10, 10)>
```

In this example using `s.T` leads to the dataset being split into just two chunks, which for large datasets means that working with this new transposed signal becomes more difficult.

However, using `da.swapaxes` the original chunking structure is preserved, potentially making it easier to work with.

```python
s1 = s.deepcopy()
s1.data = da.swapaxes(da.swapaxes(s.data, 0, 2), 1, 3)
s1.data
dask.array<swapaxes, shape=(100, 100, 50, 50), dtype=float64, chunksize=(20, 20, 10, 10)>
```

One possibility could be to have `rechunk` as an argument in `LazySignal2D` (https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/_signals/lazy.py#L874), but I'm not certain what the default behaviour should be.

**Comment by francisco-dlp** [comment]
When transposing, HyperSpy automatically rechunks the data applying the same rules as when converting a numpy array to a dask array—roughly a chunk should contain at least 1 signal its size must be ~100MB. Your customised chunks don't stick to this rule and that's why the internal re-chunking leads to very different chunks. Is this automatic re-chunking causing an issue with real use cases?

Regardless, I think that it may be a good idea to add a ``rechunk`` argument to all methods that currently automatically rechunk the data because, not only rechunking can be an expensive operation, but also the automatic chunking may not be optimal for all purposes. Are there any more opinions on this?

**Comment by magnunor** [comment]
Adding a `rechunk` argument sounds good! I'm guessing the question is what the default should be. But adding the argument itself should be straight forward.

One example where this is an issue is with some of my pixelated STEM datasets, which consist of for example 512 x 512 x 256 x 256 pixels with 2 bytes per pixel (34 GB raw), with original chunking (16, 16, 16, 16).

So to transpose, calculate a nav image and plot it can be very slow.

```python
s_nav = s.isig[128, 128].T
s_nav.compute()  # Takes 6 seconds
s.plot(navigator=s_nav)
s_t = s.T
s_t_nav = s_t.isig[128, 128].T
s_t_nav.compute()  # Takes 10+ minutes
s_t.plot(navigator=s_t_nav)
```

Also if the chunking becomes like the example in the original post, with very large chunks in the navigation dimensions, trying to plot the signal would probably crash my computer due to lack of memory.

**Comment by francisco-dlp** [comment]
I agree that this is an issue. HyperSpy's default chunking is optimized for navigation indexing i.e. to iterate over the navigation dimensions. In your example you index the signal axes to compute a navigator, and that's an expensive operation with HyperSpy's default chunking. The original chunking of your file, on the other hand, is more flexible: although not optimal for performing any particular operation, it won't be terribly slow and/or consume too much memory in most cases either.

Another issue is that HyperSpy automatically rechunks the data before performing many operations in order to optimise the chunks to the operation. This sounds fine, but in practice it doesn't work as the rechunking is "virtual" i.e. obviously the data in the file is not rechunked, and, therefore, the original chunks must be loaded into memory to compute the virtual ones. It should be possible to perform this operation without consuming plenty of memory but still requiring plenty of time. However, in my Linux machine a lot of memory is often consumed too.

For these reasons, I think that:
* All functions that currently rechunk the data should have a ``rechunk`` keyword argument.
* The default value should be ``False``.

What do you all think?

**Comment by francisco-dlp** [comment]
I am marking this as a bug because it actually makes it impossible to perform certain operations on lazy data with HyperSpy.

**Comment by magnunor** [comment]
I think having the `rechunk=False` as default is the best way to go, at least currently since there are no universally optimal ways of setting the chunking.

However, for some calculations which does not return a lazy object, it could be useful to rechunk: for example summing the last two dimensions in a 4D dataset with size (256, 256, 256, 256) and chunking (32, 32, 32, 32), changing the chunking to (32, 32, 256, 256) could speed things up. However, it would use more memory during the computation... So it might be better to leave that to the user, and just write these tips and tricks in the user guide.

**Comment by francisco-dlp** [comment]
I have just submitted a fix for this #1995

**Comment by francisco-dlp** [comment]
Fixed in #1995
